stockroom/repository.py

```diff
@@ -1,9 +1,82 @@
+from typing import Union
 from pathlib import Path
 from hangar import Repository
-from .utils import get_stock_root
+from .utils import get_stock_root, get_current_head, set_current_head
 
 
-def init(name, email, overwrite):
+class StockRepository:
+    """
+    A StockRoom wrapper class for hangar repo operations. Every hangar repo
+    interactions that is being done through stockroom (other than stock init)
+    should go through this class. Unlike hangar Repository, this class constructor
+    assumes the hangar repo is already initialized. Three class variables keep the
+    reference required for all the storages to interact with hangar repository.
+        1. Hangar repo object
+        2. Hangar repository root
+        3. Write checkout
+    Since these are class variables, they are not re-created on each initialization
+    and hence we can make sure at most, there will be only one writer checkout object
+    active at any point in time (Hangar will make sure this assertion is True anyways)
+    """
+    _write_checkout = None
+    _root: Union[Path, None] = None
+    _hangar_repo = None
+
+    def __init__(self):
+        self._setup()
+
+    @classmethod
+    def _setup(cls):
+        """
```

Comment:
This seems like an abuse of `@classmethod`, if there aren't alternative constructors, you don't need a `classmethod` for setup, you just need to run the initialization steps.

Reply:
This method updates the class variables and without making this a classmethod, we'll end up updating `StockRepository.attribute` instead of `cls.attribute`

Reply:
Ok. I'm surely missing something here, but why is that a problem?  

I think the following design solves pretty much all of your problems (and enables a few other options as well)

1. Make `_root` and `_hangar_repo` instance attributes of a `StockRepository` class (and remove `_write_checkout` attributes entirely. 
2. Use `@classmethod` as intended (as a real constructor), which allow multiple constructors based on which path to search (`cwd` or some user supplied `path`). Initialize and return an `StockRepository` instance using only these constructor methods in calling functions.
3. Replace `@classmethod checkout` method with `contextlib.contextmanager` to setup repo access ephemerally as needed in each `StockRepository` instance. Upon completion of the method block, the stdlib ensures that the `hangar.FooCheckout` (`ReaderCheckout` or `WriterCheckout`) are closed appropriatly.
4. Optional, remove the `_teardown` method since it is only used during testing. make this a test fixture function. 

Sample implementation of the new `StockRepository` class below:
 
```python
from contextlib import contextmanager


class StockRepository:
    """
    A StockRoom wrapper class for hangar repo operations. Every hangar repo
    interactions that is being done through stockroom (other than stock init)
    should go through this class. Unlike hangar Repository, this class constructor
    assumes the hangar repo is already initialized. Three class variables keep the
    reference required for all the storages to interact with hangar repository.
        1. Hangar repo object
        2. Hangar repository root
        3. Write checkout
    Since these are class variables, they are not re-created on each initialization
    and hence we can make sure at most, there will be only one writer checkout object
    active at any point in time (Hangar will make sure this assertion is True anyways)
    """

    def __init__(self, root, hangar_repo):
        self._root: os.PathLike = root
        self._hangar_repo: Repository = hangar_repo

    @property
    def stockroot(self):
        return self._root

    @contextmanager
    def checkout(self, write=False):
        """
        An api similar to hangar checkout but creates the checkout object using the
        commit hash from stock file instead of user supplying one. This enalbes users
        to rely on git checkout for hangar checkout as well.
        :param write: bool, write enabled checkout or not
        """
        try:
            if write:
                if self._hangar_repo.writer_lock_held:
                    raise PermissionError("Another write operation is in progress. "
                                        "Could not acquire the lock")
                co = self._hangar_repo.checkout(write=True)
            else:
                head_commit = get_current_head(self._root)
                co = self._hangar_repo.checkout(commit=head_commit)
            yield co
        finally:
            co.close()

    def _teardown(self):
        self._hangar_repo._env._close_environments()
        self._root = None
        self._hangar_repo = None

    @classmethod
    def _from_cwd(cls):
        """
        Setup the stock repository object by assuming the current working directory as
        the point of operation. It doesn't require user to pass the path but used the
        ``cwd``. This function creates the hangar repo object only if it is not creates
        already. It also assumes that the hangar repo is already initialized and expect
        the presence of stock file
        """
        cwd = Path.cwd()
        root = get_stock_root(cwd)
        if not root:
            raise FileNotFoundError("Could not find stock file. "
                                    "Did you forget to `stock init`?")
        hangar_repo = Repository(root, exists=True)  # hangar raises error stating not initialized
        return cls(root, hangar_repo)

    @classmethod
    def _from_path(cls, path: os.PathLike):
        root = get_stock_root(path)
        if not root:
            raise FileNotFoundError("Could not find stock file. "
                                    "Did you forget to `stock init`?")
        hangar_repo = Repository(root, exists=True)  # hangar raises error stating not initialized
        return cls(root, hangar_repo)
```

To use in calling classes, an example of the converted `GenericStore` class is below

```python
class GenericStore:
    """
    A generic data storage class that does the functionality of
    hangar metadata
    """
    def __init__(self, write):
        self._write = write
        self._repo = StockRepository._from_cwd()  # or ._from_path(foo) if desired

    def save(self, key, value):
        with self._repo.checkout(write=True) as co:
            co.metadata[parser.generic_metakey(key)] = value

    def load(self, key):
        with self._repo.checkout() as co:
            value = co.metadata[parser.generic_metakey(key)]
        return value


def genericstore(write=False):
    return GenericStore(write=write)
```

This should take care of all of your needs?

Reply:
I can see the benefits of keeping those as class variables out weight the ugliness it could cause. Making the change to convert them as instance variables. But not sure whether I should make the context manager on `checkout` especially if you decide to move `co.close` inside the hangar `checkout`'s `__exit__`. I'll use a `finally` block for now.
Also, the only change between different constructors is to find the path and hence keeping that as a check inside `__init__`. We can always revisit this later, since it doesn't really change any user facing API